libnavigation-base/src/main/java/com/mapbox/navigation/base/route/model/RouteLegProgressNavigation.kt

```diff
@@ -0,0 +1,288 @@
+                apply { this.distanceTraveled = distanceTraveled }
+
+        fun distanceRemaining(distanceRemaining: Double) =
+                apply { this.distanceRemaining = distanceRemaining }
+
+        fun durationRemaining(durationRemaining: Double) =
+                apply { this.durationRemaining = durationRemaining }
+
+        fun fractionTraveled(fractionTraveled: Float) =
+                apply { this.fractionTraveled = fractionTraveled }
+
+        fun previousStep(previousStep: LegStepNavigation) = apply { this.previousStep = previousStep }
+
+        fun currentStep(currentStep: LegStepNavigation) = apply { this.currentStep = currentStep }
+
+        fun upComingStep(upComingStep: LegStepNavigation) = apply { this.upComingStep = upComingStep }
+
+        fun followOnStep(followOnStep: LegStepNavigation) = apply { this.followOnStep = followOnStep }
+
+        fun currentStepProgress(currentStepProgress: RouteStepProgressNavigation) =
+                apply { this.currentStepProgress = currentStepProgress }
+
+        fun currentStepPoints(currentStepPoints: List<Point>?) =
+                apply { this.currentStepPoints = currentStepPoints }
+
+        fun upcomingStepPoints(upcomingStepPoints: List<Point>?) =
+                apply { this.upcomingStepPoints = upcomingStepPoints }
+
+        fun routeLeg(routeLeg: RouteLegsNavigation) = apply { this.routeLeg = routeLeg }
+
+        fun stepDistanceRemaining(stepDistanceRemaining: Double) =
+                apply { this.stepDistanceRemaining = stepDistanceRemaining }
+
+        private fun validate() {
+            var missing = ""
+            if (this.stepIndex == null) {
+                missing += " stepIndex"
+            }
+            if (this.distanceTraveled == null) {
+                missing += " distanceTraveled"
+            }
+            if (this.distanceRemaining == null) {
+                missing += " distanceRemaining"
+            }
+            if (this.durationRemaining == null) {
+                missing += " durationRemaining"
+            }
+            if (this.fractionTraveled == 0f) {
+                missing += " fractionTraveled"
+            }
+            if (this.currentStep == null) {
+                missing += " currentStep"
+            }
+            if (this.currentStepProgress == null) {
+                missing += " currentStepProgress"
+            }
+            if (this.stepDistanceRemaining == null) {
+                missing += " stepDistanceRemaining"
+            }
+            check(missing.isEmpty()) { "Missing required properties: $missing" }
+        }
+
+        fun build(): RouteLegProgressNavigation {
+            distanceTraveled = distanceTraveled()
+            fractionTraveled = fractionTraveled()
+            previousStep = previousStep()
+            upComingStep = upComingStep()
+            followOnStep = followOnStep()
+            currentStepProgress = RouteStepProgressNavigation.Builder()
```

Comment:
Same as above, those overwrite whatever would've been set using a setter. We should do one or the other.